### PR TITLE
Support read1() method in HTTPResponse

### LINKED ISSUE #2123: When iterating over a `stream`ed response with amt=None, entire response is returned instead of streaming
### Subject

While using requests, I noticed that passing `chunk_size=None` into iter_content returns the entire response when it should yield the streamed data as it arrives.

I've tracked this down to the `read(` method of `HTTPResponse`, and have a fix which simply changes this method to call `read1` instead of `read` on the underlying `_fp` when `amt=None`.

I'm happy to prepare a PR with this fix if that would be helpful (I have not written a proper test yet).

### Environment

OS macOS-10.14.6-x86_64-i386-64bit
Python 3.9.0
urllib3 1.26.2

### Steps to Reproduce

I have a reproducible example if this in the related requests issue: https://github.com/psf/requests/issues/5536

### Expected Behavior

Streamed response data should be yielded as it arrives.

### Actual Behavior

Entire response is returned.

### PR DESCRIPTION
**Pull request by smason**
Fix suggested in #2125 but I actually wanted to get #2123 fixed and this seems like a reasonable starting point.

Presumably needs more tests as I'm not doing much on that front at the moment.  But didn't want to do too much before getting some feedback.

### CODE REVIEW

File: src/urllib3/response.py [hunk 1]
```diff
@@ -876,39 +898,72 @@ def read(
...
+        self._init_decoder()
+        if decode_content is None:
+            decode_content = self.decode_content
+
+        # seems unlikely that 32MiB will ever be available in the OS buffer
+        if amt is None:
+            amt = 2**25
+
+        if len(self._decoded_buffer) >= amt:
+            return self._decoded_buffer.get(amt)
```
Comment (illia-v) [inline src/urllib3/response.py:940]: One `SSL_read` call [can return 16kB](https://www.openssl.org/docs/man3.2/man3/SSL_read.html#:~:text=The%20data%20are%20received%20in%20records%20(with%20a%20maximum%20record%20size%20of%2016kB).) of data; a small experiment shows that this amount may be decompressed to much more than 32MiB.
```sh
$ dd if=/dev/zero bs=64MiB count=1 | tr '\000' '1'  > 64MiB.txt
1+0 records in
1+0 records out
67108864 bytes (67 MB, 64 MiB) copied, 0.138876 s, 483 MB/s
$ du -sh 64MiB.txt   
64M	64MiB.txt
$ brotli 64MiB.txt
$ du -sh 64MiB.txt.br 
4.0K	64MiB.txt.br
```

Is there a reason not to return everything from `self._decoded_buffer` and make no `self._raw_read` calls when `self._decoded_buffer` is not empty?

File: src/urllib3/response.py [hunk 2]
```diff
@@ -767,6 +779,10 @@ def _fp_read(self, amt: int | None = None) -> bytes:
           * CPython < 3.10 only when `amt` does not fit 32-bit int.
         """
         assert self._fp
+        max_chunk_amt = 2**28
+        if one_read:
+            amt = max_chunk_amt if amt is None else min(amt, max_chunk_amt)
+            return self._fp.read1(amt)
```
Comment (illia-v) [inline src/urllib3/response.py:785]: It's not obvious why you have to choose a default amount and limit it to `2**28`.

I'd expect:
* something like this to be used normally:
https://github.com/urllib3/urllib3/blob/db7d96b7cc2ddc184e459301956ff203ccd91ec6/src/urllib3/response.py#L811-L812

* and only if this condition is true, the amount to be limited to `c_int_max`
https://github.com/urllib3/urllib3/blob/db7d96b7cc2ddc184e459301956ff203ccd91ec6/src/urllib3/response.py#L787-L790

File: src/urllib3/response.py [hunk 3]
```diff
@@ -876,39 +898,72 @@ def read(
...
+        if not data and len(self._decoded_buffer) == 0:
+            return data
+
+        # do not waste memory on buffer when not decoding
+        if not decode_content:
+            if self._has_decoded_content:
+                raise RuntimeError(
+                    "Calling read(decode_content=False) is not supported after "
+                    "read(decode_content=True) was called."
+                )
```
Comment (illia-v) [inline src/urllib3/response.py:952]: The method name in the error will need to be changed

**Review by illia-v — COMMENTED** [review]
Thanks for taking this on

File: src/urllib3/response.py (hunk 1 above)
Comment (smason) [reply src/urllib3/response.py:940]: That's an impressive expansion factor!  My use case didn't want enormous blocks of memory being pushed around unnecessarily, but don't think I thought it through.  They'd still exist in this case, just buffered (i.e. in `_decoded_buffer`), and hence it makes sense to just return them here when a caller asks for them.

Will update the code to actually return everything when it was requested!

Am still a bit worried about running out of memory (i.e. attempting to allocate very large `bytes` buffers for returning things like this), are there other bits of code you know of that deal with this that you know of?

File: src/urllib3/response.py (hunk 2 above)
Comment (smason) [reply src/urllib3/response.py:785]: I just wanted to do something relatively quick to code initially because I wasn't sure how likely this change was to be accepted...

The reference to `StringIO` seems weird to me as well; the types indicate that the method returns bytes, but that's not what you'd get from a `StringIO`.  I'm also wondering whether that check should be moved out into a top-method somewhere so the check wouldn't need to be done on every read.

**Comment by smason** [comment]
@illia-v thanks for the review!

As per the initial comment I wanted to get `stream(None)` returning sensible sized blocks, I presume I should do that in a separate pull-request when this one gets accepted?

**Comment by smason** [comment]
I've noticed that [`BufferedIOBase.read1`](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read1) (and `read`) are specified to use `size=-1` to indicate that "an arbitrary number of bytes are returned" whereas urllib3 uses `None` as the default value.

Is it worth changing this around at some point? that value gets propagated to the underlying stream and it feels correct to use the documented value for getting an arbitrary number of bytes.

File: src/urllib3/response.py [hunk 4]
```diff
@@ -811,7 +811,9 @@ def _fp_read(
                 return self._fp.read1(size)
         elif (
             (size > c_int_max)
-            or (self.length_remaining and self.length_remaining > c_int_max)
+            or (
+                size < 0 and self.length_remaining and self.length_remaining > c_int_max
+            )
```
Comment (smason) [inline src/urllib3/response.py:816]: Interested to know opinions about this deliberate change in semantics.

I feel as though the code should change to be doing the same chunked reading workaround when `length_remaining is None`.

**Comment by smason** [comment]
@illia-v I think I've addressed everything you noted, please let me know if I've missed anything!

**Comment by sethmlarson** [comment]
Thanks @smason! There's a lot happening in this one, would like to get some eyes on the one too.

**Comment by smason** [comment]
@sethmlarson when I started I thought it would be an easy change but with all the fiddly edge cases it became easier to do a bit of a refactor

let me know if it would help to split things out somehow, e.g. rework the commits into more logical chunks, or maybe a few smaller pull-requests?

File: test/test_response.py [hunk 5]
```diff
@@ -35,6 +35,8 @@
 from urllib3.util.response import is_fp_closed
 from urllib3.util.retry import RequestHistory, Retry
 
+OptInt: typing.TypeAlias = int | None
```
Comment (illia-v) [inline test/test_response.py:38]: Unlike other typing stuff not evaluated because of `from __future__ import annotations`, `int | None` here is so tests in Python 3.8 and 3.9 fail

**Review by illia-v — CHANGES_REQUESTED** [review]
> I've noticed that [`BufferedIOBase.read1`](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read1) (and `read`) are specified to use `size=-1` to indicate that "an arbitrary number of bytes are returned" whereas urllib3 uses `None` as the default value.
> 
> Is it worth changing this around at some point? that value gets propagated to the underlying stream and it feels correct to use the documented value for getting an arbitrary number of bytes.

It makes sense to treat -1 the same as `None` in `read` and `read1`. If this is done, setting `-1` as a default shouldn't be a breaking change and may be accepted.
I can see you made changes toward this in `_fp_read`, but `read` still treats -1 and `None` differently when it decides whether to cache content, for example.
I'd suggest keeping the previous behavior of `_fp_read` when `amt` is -1, and moving the changes to a separate future PR.

> As per the initial comment I wanted to get `stream(None)` returning sensible sized blocks, I presume I should do that in a separate pull-request when this one gets accepted?

Let's discuss possible options in a separate issue.

> let me know if it would help to split things out somehow, e.g. rework the commits into more logical chunks, or maybe a few smaller pull-requests?

Please split changes unnecessary for adding `read1` (like refactoring of `_fp_read`, changes to `read`, and type hints in existing tests) into separate small PRs, it looks like some of them are more than refactoring so they may need more attention and separate changelog entries.

**Comment by illia-v** [comment]
> As per the initial comment I wanted to get `stream(None)` returning sensible sized blocks, I presume I should do that in a separate pull-request when this one gets accepted?

Yes, please submit the change to `stream` separately.

If we use `read1` during streaming, won't it just work to pass a sensible size of a block instead of `None` as `amt`?
BTW, `amt` of [`HTTPResponse.stream`](https://urllib3.readthedocs.io/en/2.1.0/reference/urllib3.response.html#urllib3.response.HTTPResponse.stream) defaults to 65536.

**Comment by smason** [comment]
> > As per the initial comment I wanted to get `stream(None)` returning sensible sized blocks
> 
> If we use `read1` during streaming, won't it just work to pass a sensible size of a block instead of `None` as `amt`? BTW, `amt` of [`HTTPResponse.stream`](https://urllib3.readthedocs.io/en/2.1.0/reference/urllib3.response.html#urllib3.response.HTTPResponse.stream) defaults to 65536.

Yup, using `read1` is what I wanted to do.  Notably it would be able to yield byte sized chunks from endpoints like: https://httpbin.org/drip?duration=2&numbytes=8 which currently block until the stream closes before returning anything.

Having it default to 64KiB chunks would likely remain a good default; it's possible to get slightly higher performance with larger chunks but it's often not much beyond that point.

**Comment by smason** [comment]
> > I've noticed that [`BufferedIOBase.read1`](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read1) (and `read`) are specified to use `size=-1` to indicate that "an arbitrary number of bytes are returned" whereas urllib3 uses `None` as the default value.
> > Is it worth changing this around at some point? that value gets propagated to the underlying stream and it feels correct to use the documented value for getting an arbitrary number of bytes.
> 
> It makes sense to treat -1 the same as `None` in `read` and `read1`. If this is done, setting `-1` as a default shouldn't be a breaking change and may be accepted. I can see you made changes toward this in `_fp_read`, but `read` still treats -1 and `None` differently when it decides whether to cache content, for example. I'd suggest keeping the previous behavior of `_fp_read` when `amt` is -1, and moving the changes to a separate future PR.

The conditionals were getting too complicated for me to have confidence in with the prior code.  But thinking about it again it is a breaking change and hence it makes sense to keep this separate.

**Comment by illia-v** [comment]
Also, I forget to mention that I noticed that `read1` makes a `_raw_read` call even if `self._decoded_buffer` contains some bytes.
`BufferedIOBase.read1` [documents](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read1) that it reads with at most one call, so making 0 calls is allowed. What do you think about skipping `_raw_read` if `len(self._decoded_buffer) >= amt` or even always when the buffer is not empty?

**Comment by smason** [comment]
> The conditionals were getting too complicated for me to have confidence in with the prior code. But thinking about it again it is a breaking change and hence it makes sense to keep this separate.

Been trying to do this sensibly, but `_fp_read` in HEAD:

https://github.com/urllib3/urllib3/blob/4dd4a824e60ba3bb81f319cf2b8e320aa8a87dc6/src/urllib3/response.py#L783-L793

looks broken for `amt=-1`. It'll just keep trying to read the entire stream ignoring the `max_chunk_amt` constraint.

Is that actually correct?  maybe it would make sense to fix that as an initial pull-request?  The aim would be to tidy up the handling of negative vs `None` `amt`s.

**Comment by illia-v** [comment]
> Been trying to do this sensibly, but `_fp_read` in HEAD looks broken for `amt=-1`

It just doesn't expect that `-1` can be passed and neither `read` does. The response class implements `io.IOBase` which doesn't define a `read` method, so adding support for `-1` is a new feature not a bug fix 😁

If possible, let's finish this firstly keeping the current `read` logic when `-1` is passed and using the same for `read1`.

**Comment by illia-v** [comment]
> It'll just keep trying to read the entire stream ignoring the `max_chunk_amt` constraint.

You can ignore `max_chunk_amt` for `read1`, but conditions related to `c_int_max` have to be taken into account.

**Comment by illia-v** [comment]
Actually, looking at [the source code](https://github.com/python/cpython/blob/b11c443bb2ebfdd009e43ff208fa6324b658d15d/Lib/http/client.py#L456-L494) of CPython's `http.client.HTTPResponse.read`, `-1` is not expected to be passed there too.

`self._fp` is usually an instance of `http.client.HTTPResponse`.

**Comment by smason** [comment]
> > Been trying to do this sensibly, but `_fp_read` in HEAD looks broken for `amt=-1`
> 
> It just doesn't expect that `-1` can be passed and neither `read` does. The response class implements `io.IOBase` which doesn't define a `read` method, so adding support for `-1` is a new feature not a bug fix 😁

Huh, good point, that makes things easier!  I will ignore handling negative lengths for now as they seem broken in `http.client`.

> Actually, looking at [the source code](https://github.com/python/cpython/blob/b11c443bb2ebfdd009e43ff208fa6324b658d15d/Lib/http/client.py#L456-L494) of CPython's `http.client.HTTPResponse.read`, `-1` is not expected to be passed there too.

Yup, passing `-1` will read until the end of the stream, bypassing the [check](https://github.com/python/cpython/blob/b11c443bb2ebfdd009e43ff208fa6324b658d15d/Lib/http/client.py#L469-L471) that `amt <= self.length`.  That's assuming that `http.client.HTTPResponse`'s `fp` is the expected `BufferedReader` that comes from a socket's `makefile`.

Can't find much in CPython's Github issues, about this.  There are some mentions of the io API and negative values in https://github.com/python/cpython/issues/67403 but nothing much closer.  The following MWE just hangs for me:
```python
con = http.client.HTTPConnection("httpbin.org")
con.request("GET", "/")
resp = con.getresponse()
assert resp.headers["connection"] == "keep-alive"
print(resp.read(-1))
```
which feels wrong, or am I missing something?  Is it worth raising an issue about this?

**Comment by smason** [comment]
> > which feels wrong, or am I missing something?
> 
> The example prints HTML for me but with a significant delay, this feels wrong to me too. `http.client.HTTPResponse` [is said](https://docs.python.org/3/library/http.client.html#httpresponse-objects) to implement `io.BufferedIOBase` since Python 3.5, so I'd expect it to treat `None` and a negative number as the `read` argument the same.

Glad it's not just me misinterpreting APIs!  Have raised https://github.com/python/cpython/issues/112064 about this.

File: src/urllib3/response.py [hunk 6]
```diff
@@ -768,7 +796,14 @@ def _fp_read(self, amt: int | None = None) -> bytes:
         """
         assert self._fp
         c_int_max = 2**31 - 1
-        if (
+        if one_read:
+            # StringIO doesn't like amt=None
+            if amt is None:
+                return self._fp.read1()
+            if amt > c_int_max and (util.IS_PYOPENSSL or sys.version_info < (3, 10)):
+                return self._fp.read1(c_int_max)
```
Comment (illia-v) [inline src/urllib3/response.py:804]: I believe there is a possibility of an overflow error here because the `(self.length_remaining and self.length_remaining > c_int_max)` part of the following condition is skipped for `one_read`.
https://github.com/urllib3/urllib3/blob/4dd4a824e60ba3bb81f319cf2b8e320aa8a87dc6/src/urllib3/response.py#L771-L774

When `amt` is `None` and the content length is known, `http.client.HTTPResponse.read1` will [try to read](https://github.com/python/cpython/blob/9302f05f9af07332c414b3c19003efd1b1763cf3/Lib/http/client.py#L654-L655) the length amount of bytes. This can lead to an overflow error as described in the docstring of `_fp_read`.

Ideally (and probably required for 100% coverage), a test similar to existing `test_requesting_large_resources_via_ssl` has to check possible cases for `read1`. The problem with the existing test is that it causes timeouts in CI. Adding a similar one will worsen the situation. Although, a test for `read1` should be quicker because it will not read all 2 GiB.

File: src/urllib3/response.py [hunk 7]
```diff
@@ -909,6 +946,41 @@ def read(
...
+        if decode_content is None:
+            decode_content = self.decode_content
+        if not decode_content and self._has_decoded_content:
+            raise RuntimeError(
+                "Calling read1(decode_content=False) is not supported after "
+                "read1(decode_content=True) was called."
+            )
+
+        if amt is not None and amt < len(self._decoded_buffer):
+            return self._decoded_buffer.get(amt)
```
Comment (illia-v) [inline src/urllib3/response.py:964]: > Read with at most one underlying system call.  If at least one byte is buffered, return that instead.

This is from the [docstring](https://github.com/python/cpython/blob/9302f05f9af07332c414b3c19003efd1b1763cf3/Lib/http/client.py#L647-L648) of `http.client.HTTPResponse.read1`. I would follow the last sentence in urllib3 too.

So, if `self._decoded_buffer` is not empty, make no `self._raw_read` calls and:
* when `amt` is a number, return up to `amt` bytes from the buffer;
* when `amt` is `None`, return everything from the buffer.

@sethmlarson @pquentin what do you think about this?

**Review by illia-v — COMMENTED** [review]
> I've squashed all my WIP changes into a single one that's just about adding read1 support.

Thanks, this is really helpful!

File: src/urllib3/response.py (hunk 7 above)
Comment (smason) [reply src/urllib3/response.py]: Avoiding calling into the kernel makes sense, I just copied the semantics of `read` which obviously don't apply here.  Have changed that condition to return early when possible.

Be interesting to see what other people think!

**Review by illia-v — COMMENTED** [review]
This looks pretty good with the latest changes!

I merged the main branch with some updates fixing CI timeouts. Could you please add more tests to fix [the coverage misses](https://github.com/urllib3/urllib3/actions/runs/6876950948/job/18705065000)?
Also, please add a test similar to `test_requesting_large_resources_via_ssl` for `read1`,  and check the docs building failure in CI.

File: src/urllib3/response.py [hunk 8]
```diff
@@ -909,6 +948,57 @@ def read(
...
+                )
+            if len(self._decoded_buffer) > 0:
+                if amt is None:
+                    return self._decoded_buffer.get_all()
+                return self._decoded_buffer.get(amt)
+        if amt == 0:
+            return b""
+
+        data = self._raw_read(amt, one_read=True)
+        if (not decode_content) or (not data and len(self._decoded_buffer) == 0):
```
[the remark this answers is not in the document]
Comment (smason) [reply src/urllib3/response.py:985]: The decoder still needs to be flushed in this case doesn't it?  i.e. not sure if it's valid to bypass the `decode_content` check

**Comment by smason** [comment]
@illia-v Have rebased and applied those two fixes

Will look at writing some tests to get the coverage back up

File: src/urllib3/response.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (illia-v) [reply src/urllib3/response.py]: It looks like `read` doesn't get to flushing the decoder when there is no data

[the remark this answers is not in the document]
Comment (smason) [reply src/urllib3/response.py]: > It looks like `read` doesn't get to flushing the decoder when there is no data

The code you linked to is in `read1`, are you referring to the original `read` method here?

I'm asking because I don't see any issue with my logic in `read1`.  I've just noticed it still contains some redundancies due to me basing it on the `read` implemention (two very similar looking calls to `self._decode()` and `self._decoded_buffer.put()`), but I don't think that affects correctness.

**Comment by smason** [comment]
@illia-v coverage seems to be back up to 100% now

not sure why the docs build is giving:

> WARNING: py:meth reference target not found: http.client.HTTPResponse.read1

I can't find any special handling for `http.client.HTTPResponse.read` and yet it finds that method fine

**Comment by illia-v** [comment]
> @illia-v coverage seems to be back up to 100% now

Awesome ✨

> not sure why the docs build is giving:
> 
> > WARNING: py:meth reference target not found: http.client.HTTPResponse.read1
> 
> I can't find any special handling for `http.client.HTTPResponse.read` and yet it finds that method fine

I am not sure why too, one assumption is that it happens because `http.client.HTTPResponse.read1` is not documented by CPython https://docs.python.org/3/library/http.client.html#httpresponse-objects

File: src/urllib3/response.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (illia-v) [reply src/urllib3/response.py]: > > It looks like `read` doesn't get to flushing the decoder when there is no data
> 
> The code you linked to is in `read1`, are you referring to the original `read` method here?

Sorry for confusing you. I meant that the `read` method on the following lines returns an empty bytes object regardless of `decode_content`.
https://github.com/urllib3/urllib3/blob/bc41d52707c0f7717364bd571adc2d3f353ad60e/src/urllib3/response.py#L919-L920
It's like an early exit. So I assumed an `if not data: return data` condition will copy the behavior from `read` to `read1`. Do you see any issue in doing so?

[the remark this answers is not in the document]
Comment (smason) [reply src/urllib3/response.py]: it looks like a bug to me...

`git blame` doesn't suggest anything obvious, https://github.com/urllib3/urllib3/commit/c35033f6cc54106ca66ef2d48a9e3564d4fb0e07 is when `read` was the last changed and was quite a large change.

maybe there's a test for that specific behavior you know of?

[the remark this answers is not in the document]
Comment (illia-v) [reply src/urllib3/response.py]: I commented out the mentioned `read` lines, and many tests failed because `self._fp` was `None`, so `self._raw_read` returned `None`, and then joining `None` with some bytes was attempted

It's possible to reproduce a similar error by running `HTTPResponse().read1()`

[the remark this answers is not in the document]
Comment (smason) [reply src/urllib3/response.py]: Why does `_raw_read` return `None`?! the return type is just `bytes`!

Maybe it should raise an exception when the stream has been closed?  Presume this is due to the stream getting implicitly closed when reading, but that feels as wrong.  Probably a another separate issue.

I also note that the whole content decoding feels backwards to me, why does `_init_decoder` keep getting called on every read which then pulls the header string out every time.  Surely that can be done once and then remembered — the headers shouldn't change should they?

**Comment by smason** [comment]
It occurred to me that `read` should be using `read1` when a decoder is in use, i.e. doing something like `_raw_read(65536, one_read=True)` rather than just passing `amt` down the chain.

The user might have requested a specific number of bytes, but because of compression it likely won't take a read of that size from the socket to fulfill the request.

File: src/urllib3/response.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (illia-v) [reply src/urllib3/response.py]: I agree that there is a room for tidying up this part of urllib3
Regarding `_raw_read`: returning `None` is a pity, maybe that's a consequence of `self._fp` defaulting to `None` in some cases like in tests, so that should be changed

**Comment by illia-v** [comment]
> It occurred to me that `read` should be using `read1` when a decoder is in use, i.e. doing something like `_raw_read(65536, one_read=True)` rather than just passing `amt` down the chain.
> 
> The user might have requested a specific number of bytes, but because of compression it likely won't take a read of that size from the socket to fulfill the request.

I'd expect a user to choose `amt` based on `Content-Length` that provides a size of an encoded (possibly compressed) body

File: src/urllib3/response.py [hunk 9]
```diff
@@ -280,6 +280,22 @@ def get(self, n: int) -> bytes:
...
+        if not buffer:
+            assert self._size == 0
+            return b""
+        if len(buffer) == 1:
+            result = buffer.pop()
+        else:
+            # BytesIO.writelines says "newlines are not added"
+            with io.BytesIO() as fd:
+                fd.writelines(buffer)
+                result = fd.getvalue()
```
Comment (illia-v) [inline src/urllib3/response.py:294]: Can this be simplified?
```suggestion
            result = b"".join(buffer)
```

**Review by illia-v — DISMISSED** [review]
LGTM, just one small comment

@sethmlarson @pquentin do you want to take a look at the PR before we merge it?

**Comment by pquentin** [comment]
@illia-v I fully trust your judgment. Thanks!

**Review by sethmlarson — COMMENTED** [review]
High-level question, how is `read1()` different than `read()`'s implementation except for passing `read1=True` into `_raw_read()`? If that is a potential implementation path (even if it's not as efficient due to not using `get_all()`) can that be considered?

I ask this because our HTTPResponse data reading is already so complicated, having two parallel and slightly different implementations makes me worried for future maintenance.

File: src/urllib3/response.py [hunk 10]
```diff
@@ -770,8 +798,14 @@ def _fp_read(self, amt: int | None = None) -> bytes:
         c_int_max = 2**31 - 1
         if (
             (amt and amt > c_int_max)
-            or (self.length_remaining and self.length_remaining > c_int_max)
+            or (
+                amt is None
```
[the remark this answers is not in the document]
Comment (smason) [reply src/urllib3/response.py:802]: @sethmlarson `read1` is a standard method provided by `io.BufferedReader`s which `http.client.HTTPResponse` also provides

this PR is mostly about getting `read1` supported, there are various cleanups that make sense once it's in there.  e.g. with compressed streams it doesn't make sense to me that `amt` is passed to `read` on the stream.  the decompressed data might be much larger and hence using `read1` would mean that the caller might get a response much earlier

> I'm not sure why `amt is None` was added here? The new branch `if one_read` doesn't use `amt`.

without it this "slow path" would get used for large responses even if `amt` is small

File: src/urllib3/response.py (hunk 9 above)
Comment (smason) [reply src/urllib3/response.py:294]: not sure why that didn't occur to me! think I was basing it in `get` and didn't notice that's all it was doing in the end

**Comment by smason** [comment]
> > It occurred to me that `read` should be using `read1` when a decoder is in use, i.e. doing something like `_raw_read(65536, one_read=True)` rather than just passing `amt` down the chain.
> > The user might have requested a specific number of bytes, but because of compression it likely won't take a read of that size from the socket to fulfill the request.
> 
> I'd expect a user to choose `amt` based on `Content-Length` that provides a size of an encoded (possibly compressed) body

Speaking as a user of `requests` I don't think I've ever done that!  `HTTPResponse.stream` certainly doesn't do anything like that, so why would calling code (that presumably knows even less details about what's going on at the HTTP level) know what makes sense to pass here?

**Comment by smason** [comment]
I've squashed and rebased again, hopefully that will lead to a cleaner merge

File: src/urllib3/response.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (illia-v) [reply src/urllib3/response.py:799]: @sethmlarson in case urllib3 or http.client is going to `read1` 2 GiB or more and get the overflow error, the code under this if block prevents this by passing the maximum safe amount to `read1`

The condition would be more correct with the new `amt is None` check even without the `read1` addition. Not having it had no adverse effect previously, but now it's either `amt is None` has to be added here or code on lines 804–805 has to become more complex.

**Comment by illia-v** [comment]
> High-level question, how is `read1()` different than `read()`'s implementation except for passing `read1=True` into `_raw_read()`?

I can see two changes besides calling `_raw_read`:
* if there is at least one byte in `self._decoded_buffer`, `read1` unlike `read` will always return the buffered content and make no `_raw_read` calls;
* `read1` doesn't try to call `_raw_read` until `amt` is reached.

**Comment by pquentin** [comment]
> I can see two changes besides calling `_raw_read`

Should this be in a code comment?

**Comment by smason** [comment]
> > I can see two changes besides calling `_raw_read`
> 
> Should this be in a code comment?

The docstring of `read1` points to [`io.BufferedReader.read1`](https://docs.python.org/3/library/io.html#io.BufferedIOBase.read1) as guiding the semantics, i.e. performing at most one IO, and there's a comment near the top of `read1` saying "try and respond without going to the network".  I've tried to keep the code readable and saying much more would just seem to duplicate the code in comments, which is often counter productive in my experience.

I've tried to stay away from changing `read` as the changes were quickly getting out of hand.

**Review by illia-v — APPROVED** [review]

**Comment by smason** [comment]
thanks for merging!  :tada:
